sdks/java/core/src/main/java/org/apache/beam/sdk/fn/stream/DirectStreamObserver.java

```diff
@@ -53,7 +53,7 @@ public final class DirectStreamObserver<T> implements StreamObserver<T> {
   private final int maxMessagesBeforeCheck;
 
   private final Object lock = new Object();
-  private int numMessages = -1;
```

Comment:
This seems ok, but I'm not clear why this is less racy? 
Why is initializing to 0 different from initializing to -1?

can this be annotated with GuardedBy?